### PR TITLE
Catch exception in ~WriteBufferFromS3

### PR DESCRIPTION
**Pull request by vdimir**
Changelog category (leave one):

- Not for changelog (changelog entry is not required)

Function `finalize` in `~WriteBufferFromS3` may throw an exception

~https://s3.amazonaws.com/clickhouse-test-reports/34951/96390a9263cb5af3d6e42a84988239c9ae87ce32/stress_test__debug__actions_.html~

https://s3.amazonaws.com/clickhouse-test-reports/34951/96390a9263cb5af3d6e42a84988239c9ae87ce32/stress_test__undefined__actions_/clickhouse-server.err.log

### CODE REVIEW

File: src/IO/WriteBufferFromS3.cpp [hunk 1]
```diff
@@ -116,7 +116,14 @@ void WriteBufferFromS3::allocateBuffer()
 
 WriteBufferFromS3::~WriteBufferFromS3()
 {
-    finalize();
```
Comment (azat) [inline src/IO/WriteBufferFromS3.cpp:119]: Maybe it is better to put `finalize` call into caller to avoid suppressing errors?

**Comment by azat** [comment]
>https://s3.amazonaws.com/clickhouse-test-reports/34951/96390a9263cb5af3d6e42a84988239c9ae87ce32/stress_test__debug__actions_.html

By the way here I see only jemalloc exception, not uncaught exception.
Am I missing something?

**Comment by vdimir** [comment]
@azat 
Oops, I've attached wrong report 

https://s3.amazonaws.com/clickhouse-test-reports/34951/96390a9263cb5af3d6e42a84988239c9ae87ce32/stress_test__undefined__actions_/clickhouse-server.err.log

**Comment by azat** [comment]
>Oops, I've attached wrong report

At least I've finally figured out why the assertion in jemalloc was hitting (hopefully).

>
https://s3.amazonaws.com/clickhouse-test-reports/34951/96390a9263cb5af3d6e42a84988239c9ae87ce32/stress_test__undefined__actions_/clickhouse-server.err.log

So maybe it worth to do proper destroy in StorageS3Sink, instead of adding try/catch in dtor.

File: src/QueryPipeline/QueryPipeline.cpp [hunk 2]
```diff
@@ -528,6 +532,16 @@ void QueryPipeline::addStorageHolder(StoragePtr storage)
     resources.storage_holders.emplace_back(std::move(storage));
 }
 
+
+void QueryPipeline::cancel()
```
Comment (KochetovNicolai) [inline src/QueryPipeline/QueryPipeline.cpp:536]: Don't fully understand why this is needed ...

**Comment by vdimir** [comment]
@azat 

Still not sure what's the proper way.

We can call `cancel` or `finish` for all processors for `QueryPipeline::reset` 


https://github.com/ClickHouse/ClickHouse/blob/c48c13601dd84eee70cfd0904827ff9859076b1a/src/QueryPipeline/QueryPipeline.cpp#L533

But it would throw anyway. Also, should any retries be performed after errors from `~WriteBufferFromS3`?

**Comment by vdimir** [comment]
Let's merge current solution with `try ... catch ...` in `~WriteBufferFromS3`, proper fix can be added later

**Comment by azat** [comment]
>But it would throw anyway. Also, should any retries be performed after errors from ~WriteBufferFromS3?

Yes, but it should call `onFinish` that calls `WriteBufferFromS3::finalize` and it will not throw from dtor since `finalize` will set to `true`, so `finalize` from dtor will be no-op.

**Comment by azat** [comment]
>what happened if finalize throw excaption? Looks like there is a possibility of data loss.

Indeed, and to me it is better to abort over data loss.
That's why initially (#19451) there was no try/catch in dtors.

**Pull request merged**